src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/RecordingUploadPostHandler.java

```diff
@@ -0,0 +1,174 @@
+        this.webClient = webClient;
+        this.fs = fs;
+        this.savedRecordingsPath = savedRecordingsPath;
+    }
+
+    @Override
+    public HttpMethod httpMethod() {
+        return HttpMethod.POST;
+    }
+
+    @Override
+    public String path() {
+        return "/api/v1/recordings/:recordingName/upload";
+    }
+
+    @Override
+    public boolean isAsync() {
+        return false;
+    }
+
+    @Override
+    void handleAuthenticated(RoutingContext ctx) throws Exception {
+        String recordingName = ctx.pathParam("recordingName");
+        try {
+            URL uploadUrl = new URL(env.getEnv("GRAFANA_DATASOURCE_URL"));
+            // TODO validate URL before POST attempt
+            ResponseMessage response = doPost(recordingName, uploadUrl);
+            ctx.response().setStatusCode(response.statusCode);
+            ctx.response().setStatusMessage(response.statusMessage);
+            ctx.response().end(response.body);
+        } catch (MalformedURLException e) {
+            throw new HttpStatusException(501, e);
+        }
+    }
+
+    // FindBugs thinks the recordingPath or its properties is null somehow
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
+    private ResponseMessage doPost(String recordingName, URL uploadUrl) throws Exception {
+        Path recordingPath =
+                getRecordingPath(recordingName)
+                        .orElseThrow(() -> new HttpStatusException(404, recordingName));
+
+        MultipartForm form =
+                MultipartForm.create()
+                        .binaryFileUpload(
+                                "file",
+                                recordingName,
+                                recordingPath.toString(),
+                                HttpMimeType.OCTET_STREAM.toString());
+
+        CompletableFuture<ResponseMessage> future = new CompletableFuture<>();
+        webClient
+                .postAbs(uploadUrl.toURI().resolve("/load").normalize().toString())
+                .timeout(30_000L)
+                .sendMultipartForm(
+                        form,
+                        uploadHandler -> {
+                            if (uploadHandler.failed()) {
+                                future.completeExceptionally(uploadHandler.cause());
+                                return;
+                            }
+                            HttpResponse<Buffer> response = uploadHandler.result();
+                            future.complete(
+                                    new ResponseMessage(
+                                            response.statusCode(),
+                                            response.statusMessage(),
+                                            response.bodyAsString()));
+                        });
+        return future.get();
+    }
+
+    Optional<Path> getRecordingPath(String recordingName) throws Exception {
+        Path archivedRecording = savedRecordingsPath.resolve(recordingName);
```

Comment:
`resolve` can throw an `InvalidPathException`. If this happens, it will throw through `handleAuthenticated` as it only catches `MalformedURLException`. Would it make sense to handle it within instead to respond that the input was invalid? Or does it do that somewhere higher up?

I see test class for `TargetRecordingUploadPostHandler` but not for this one. Is that intended?

Reply:
`handleAuthenticated` is called in `AbstractAuthenticatedRequestHandler.handle()`, which will catch any exception thrown by `handleAuthenticated` and handle it appropriately. For an `InvalidPathException` or any other generally unexpected exception, it will be wrapped in an `HttpStatusException` with status code 500, which will later on be caught by the `WebServer` and turned into a failure response on the wire.

Reply:
re: missing test - nope, not intended, looks like I forgot to write one. Good catch :-) I'll add one now.

Reply:
Okay. I agree, I think it can be caught in this class and handled as a `4xx`.